### DISCUSSION TITLE
How does the automatic derivative of x w.r.t. p (sensitivity analysis) works?

### DISCUSSION THREAD

**Discussion by yakunouyang** [casadi/casadi#3131] [Q&A] [answered]
Hello 👋

I'm following the tutorial blog about [Sensitivities of parametric NLP](https://web.casadi.org/blog/nlp_sens/), understand most of it, but how does the jacobian evaluation of x w.r.t. p. works internally?

after some research on example files, I've created following script to help me understanding it, but ran into some error warm-starting the sqpmethod.

```cpp
#include <casadi/casadi.hpp>

using namespace casadi;

int main() {
  auto x = SX::sym("x");
  auto y = SX::sym("y");
  auto p = SX::sym("p");
  SXDict nlp = {{"x", SX::vertcat({x, y})},
                {"p", p},
                {"f", sq(1 - x) + p * sq(y - sq(x))},
                {"g", x + y}};
  auto solver = nlpsol("solver", "ipopt", nlp);
  DMDict solver_in;
  solver_in["x0"] = std::vector<double>{0, 1};
  solver_in["lbx"] = std::vector<double>{-10, 1.2};
  solver_in["ubx"] = std::vector<double>{10, 2};
  solver_in["lbg"] = -10;
  solver_in["ubg"] = 10;
  solver_in["p"] = 100;
  auto solver_out = solver(solver_in);
  std::cout << solver_out["x"] << std::endl;

  auto taylor_solver = nlpsol("taylor_solver", "sqpmethod", nlp);
  auto taylor = taylor_solver.factory("taylor", taylor_solver.name_in(), {"x", "jac:x:p"});
  DMDict taylor_options = solver_in;
  taylor_options["x0"] = solver_out["x"];
  taylor_options["lam_x0"] = solver_out["lam_x"];
  taylor_options["lam_g0"] = solver_out["lam_g"];

  auto tsol = taylor(taylor_options);
  std::cout << tsol["x"] << std::endl;
  std::cout << tsol["jac_x_p"] << std::endl;
  return 0;
}
```

I've acknowledged that the jacobian evaluation process requires sqpmethod to solve the orginal NLP problem again, and warm-starting the sqpmethod with solution from IPOPT is helping, so I tried to introduce `x0`, `lam_x` and `lam_g` from IPOPT  to further accelerate the sqpmethod, the sqpmethod converged with 0 iterations but the jacobian evalation function seems to fail, as shown in following output.

```
iter      objective    inf_pr    inf_du     ||d||  lg(rg) ls    info
   0   9.090831e-03  0.00e+00  1.69e-13  0.00e+00       -  0  - 
MESSAGE(sqpmethod): Convergence achieved after 0 iterations
taylor_solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
   nlp_jac_fg  |   2.00us (  2.00us)   2.62us (  2.62us)         1
        total  |  37.00us ( 37.00us)  35.92us ( 35.92us)         1
fwd1_taylor_solver  :   t_proc      (avg)   t_wall      (avg)    n_eval
             total  |  57.00us ( 57.00us)  57.00us ( 57.00us)         1
libc++abi: terminating with uncaught exception of type casadi::CasadiException: Error in Function::call for 'taylor' [MXFunction] at .../casadi/core/function.cpp:1401:
Error in Function::call for 'taylor' [MXFunction] at .../casadi/core/function.cpp:330:
.../casadi/core/function_internal.hpp:1607: Evaluation failed
```

if I comment out the `lam_x` and `lam_g` lines, the `taylor_solver` takes 1 iteration to converge and the `fwd1_taylor_solver` prints the result successfully (which seems reasonable but large problem takes more iterations).

does the jacobian evaluation function `fwd1_taylor_solver` depends on sqpmethod taking some iterations, or is my code wrong? 

besides, the `fwd1_taylor_solver` is very slow (~3s) with large NLP problem (1726 variables and 1417 constraints), is there some way to improve it?

CasADi is a great software, but understanding how it works is hard 😰. Thanks for helping!

Yakun Ouyang

**Comment by jaeandersson** [comment]
The mathematical background is explained in this paper: https://www.sciencedirect.com/science/article/pii/S2405896318327137

**Reply by yakunouyang** [reply]
> The mathematical background is explained in this paper: https://www.sciencedirect.com/science/article/pii/S2405896318327137

Thanks for quick reply! but I still don't quite understand why the problem failed after providing the `lam_x` and `lam_g` values, can you help me with that?

**Reply by brunoml5** [reply]
Also, I believe that the sensitivity should work better with the QR solver for the QP in `sqpmethod`, right? While the default QP solver for `sqpmethod` is qpOASES. So, maybe try switching to `qrsqp`, maybe it converges better.

**Reply by jaeandersson** [reply]
> > The mathematical background is explained in this paper: https://www.sciencedirect.com/science/article/pii/S2405896318327137
> 
> Thanks for quick reply! but I still don't quite understand why the problem failed after providing the `lam_x` and `lam_g` values, can you help me with that?

I don't have a quick answer to why it doesn't work in this case, without digging into the code. You might want to use the lower-level functions `taylor_solver.forward` or `taylor_solver.reverse` instead of `taylor_solver.factory`. That way you avoid having to resolve the NLP in the derivative calculations. And it might also be easier to debug.

**Comment by yakunouyang** [comment, marked as answer]
After further digging into the code and printing instructions, I found out the forward evaluation is actually stopped at [this line](https://github.com/casadi/casadi/blob/7e44a2bd28b92545b192ed91097cd288a8589961/casadi/core/nlpsol.cpp#L1028), which solves a linear equation (Eq. 13 from [1]) to acquire forward sensitivities, since the `lam_x` and `lam_g` is passed directly to `sqpmethod` and converged immediately, the forward process actually starts with IPOPT solutions, and IPOPT returns small epsilon multiplier for inactive constraints. Then I discovered issue #2660, @jgillis added `clip_inactive_lam` option to IPOPT plugin to make multiplier exactly zero, **setting `clip_inactive_lam` to true solves the problem in this post**. Besides, setting `min_lam` to a larger epsilon value > 0 seems to have same effect. Does that mean we don't need `sqpmethod` to do sensitivity analysis in CasADi?

Regarding second question (reduce fwd_solver time), after some profiling on instructions, I found that solve Eq. 13 takes up most of the time, this is because the default value for `sens_linsol` option is `qr`, change it to `ma27` significantly reduces forward solver time (from ~3s to ~5ms).

Lastly, I think the stacktrace for function evaluation can be improved, currently it just aborted with message `Evaluation failed`, which is very confusing 😢. 

Thanks all for helping, learned alot debugging!

[1] https://www.sciencedirect.com/science/article/pii/S2405896318327137
